src/common/src/mlib/vec.t.h

```diff
@@ -0,0 +1,450 @@
+ *   The first macro argument is a pointer to the element and subsequent arguments
+ *   are unspecified and reserved for future use. Elements are zero-initialized
+ *   before being passed to this macro.
+ * - Optional: Define `VecCopyElement(DstPtr, SrcPtr)` to copy data from `*SrcPtr`
+ *   to `*DstPtr`. The vector's copying function is only defined if this macro
+ *   is defined. This macro MUST evaluate to a boolean to indicate if the copy
+ *   operation succeeded. If a copy fails, then the partially copied elements
+ *   will be destroyed and the overall copy will fail.
+ *
+ *   To add a trival copying function, define `VecCopyElement` to
+ *   `VecTrivialCopyElement`.
+ *
+ * - NOTE: All of the above macros will be automatically undef'd after this file
+ *   is included.
+ *
+ * Types stored in the vector must be trivially relocatable.
+ *
+ * @copyright Copyright 2009-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <mlib/ckdint.h>
+#include <mlib/config.h>
+#include <mlib/intutil.h>
+
+#include <assert.h>  // assert
+#include <stdbool.h> // bool
+#include <stddef.h>  // size_t
+#include <stdlib.h>  // calloc, realloc, free
+#include <string.h>  // memcpy, memset
+
+// Check that the caller provided a `T` macro to be the element type
+#ifndef T
+#if defined(__clangd__) || defined(__INTELLISENSE__)
+#define T int                                // Define a type for IDE diagnostics
+#define VecCopyElement VecTrivialCopyElement // For IDE highlighting
+#else
+#error A type `T` should be defined before including this file
+#endif
+#endif
+
+#ifndef VecName
+#define VecName MLIB_PASTE(T, _vec)
+#endif
+
+#ifndef VecDestroyElement
+#define VecDestroyElement(Ptr) ((void)(Ptr))
+#endif
+
+#ifndef VecInitElement
+#define VecInitElement(Ptr) ((void)(Ptr))
+#endif
+
+#ifndef VecTrivialCopyElement
+#define VecTrivialCopyElement(DstPtr, SrcPtr) ((*(DstPtr) = *(SrcPtr)), true)
+#endif
+
+#pragma push_macro("vec_inline_spec")
+#if !defined(vec_inline_spec)
+#define vec_inline_spec static inline
+#endif
+
+// The "fn" macro just adds a qualified name to the front of a function identifier
+#pragma push_macro("fn")
+#undef fn
+#define fn(M) MLIB_PASTE_3(VecName, _, M)
+
+typedef struct VecName {
+   /**
+    * @private
+    * @brief Pointer to the first vector element, or NULL if the vector is
+    * empty.
+    *
+    * @note DO NOT MODIFY
```

Comment:
```suggestion
    * @note DO NOT DIRECTLY MODIFY THIS VALUE
```
Suggest being a bit more specific about what these notes mean.